src/types/ErrorRate.ts

```diff
@@ -0,0 +1,171 @@
+import { NA, RequestHealth } from './Health';
+import { RateHealthConfig, ToleranceConfig } from './ServerConfig';
+import { serverConfig } from '../config/ServerConfig';
+import { ThresholdStatus, getRequestErrorsStatus, RATIO_NA } from './Health';
+
+export interface ErrorRatio {
+  errorRatio: { status: ThresholdStatus; protocol?: string };
+  inboundErrorRatio: { status: ThresholdStatus; protocol?: string };
+  outboundErrorRatio: { status: ThresholdStatus; protocol?: string };
+}
+
+export interface Rate {
+  requestRate: number;
+  errorRate: number;
+  errorRatio: number;
+}
+
+export const regexProtocol = {
+  grpc: new RegExp(/^[1-9]$|^1[0-6]$/),
+  /* eslint-disable-next-line */
+  http: new RegExp(/^[4-5]\d\d$/)
+};
+
+export const checkExpr = (expr: string, value: string): boolean => {
+  if (expr !== '' && value !== '') {
+    const reg = new RegExp(expr);
+    return reg.test(value);
+  }
+  return false;
+};
+
+const getConfig = (ns: string, name: string, kind: string): RateHealthConfig | undefined => {
+  for (let rate of serverConfig.healthConfig.rate) {
+    if (checkExpr(rate.namespace, ns) && checkExpr(rate.name, name) && checkExpr(rate.kind, kind)) {
+      return rate;
+    }
+  }
+  return undefined;
+};
+
+export const calculateErrorRate = (
+  ns: string,
+  name: string,
+  kind: string,
+  requests: RequestHealth
+): { errorRatio: ErrorRatio; config: RateHealthConfig | undefined } => {
+  const conf = getConfig(ns, name, kind);
+  const inboundTolerances = conf ? conf.tolerance.filter(tol => checkExpr(tol.requestType, 'inbound')) : undefined;
```

Comment:
Hint: if I'm not wrong you can use here the new "optional chaining operator" here: ` = conf?.tolerance.filter(...)`
It's up to you :)

Reply:
Done